consensus/hotstuff/verification/common.go

```diff
@@ -66,3 +68,77 @@ func checkVotesValidity(votes []*model.Vote) error {
 
 	return nil
 }
+
+// stakingKeysAggregator is a structure that aggregates the staking
+// public keys for QC verifications.
+type stakingKeysAggregator struct {
+	lastStakingSigners map[*flow.Identity]struct{}
+	lastStakingKey     crypto.PublicKey
+	lock               sync.Mutex
+}
+
+// creates a new staking keys aggregator
+func newStakingKeysAggregator() *stakingKeysAggregator {
+	aggregator := &stakingKeysAggregator{
+		lastStakingSigners: map[*flow.Identity]struct{}{},
+		lastStakingKey:     crypto.NeutralBLSPublicKey(),
+		lock:               sync.Mutex{},
+	}
+	return aggregator
+}
+
+// aggregatedStakingKey returns the aggregated public of the input signers.
+func (s *stakingKeysAggregator) aggregatedStakingKey(signers flow.IdentityList) (crypto.PublicKey, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// this greedy algorithm assumes the signers set does not vary much from one call
```

Comment:
👍  Nice optimization

better to explain the reasoning of why the signers set doesn't vary much.

I guess it's because we currently have each consensus node having equal amount of weight. Since we require 2/3 votes, the difference is at most 1/3 signers, right? So the worse case is just to aggregate 1/3 keys + 1 aggregated key, which is always better than aggregating 2/3 keys.

Reply:
Almost right :) 
Considering equal weights, the worst case is actually aggregating 2/3 + 1 of the keys : this happens when the new list of 2/3 has only 1/3 in common with the previous list of 2/3. In this case, the new keys to add are 1/3 and the missing keys are 1/3 also. Removing takes the same time as adding, so this gives 2/3 + 1. Which is the same as aggregating 2/3 from scratch.

Reply:
Question to you btw: does a block proposer strictly stop gathering votes in the QC as soon as it reaches 67% or can it go sometimes a bit further, like to 70% or more, because it received many votes simultaneously?

Reply:
I think the VoteAggregator will stop collecting votes as soon as there are enough votes, build the QC, cache the QC and return it. So, we'll only collect as few the min amount possible to get _more_ than 2/3 voters (by stake)